src/code42cli/logger/handlers.py

```diff
@@ -0,0 +1,130 @@
+    made for TCP/TLS until the first log record is about to be transmitted.
+
+    Args:
+        hostname: The hostname of the syslog server to send log messages to.
+        port: The port of the syslog server to send log messages to.
+        protocol: The protocol over which to submit syslog messages. Accepts TCP, UDP, or TLS.
+    """
+
+    def __init__(self, hostname, port, protocol, certs):
+        self._hostname = hostname
+        self._port = port
+        self._certs = certs
+        self.address = (hostname, port)
+        logging.Handler.__init__(self)
+        self._use_insecure = protocol != ServerProtocol.TLS_TCP
+        self.socktype = _try_get_socket_type_from_protocol(protocol)
+        self.socket = None
+
+    def connect_socket(self):
+        """Call to initialize the socket. If using TCP/TLS, it will also establish the connection."""
+        if not self.socket:
+            self.socket = self._create_socket(
+                self._hostname, self._port, self._use_insecure, self._certs
+            )
+
+    def _create_socket(self, hostname, port, use_insecure, certs):
+        socket_info = self._get_socket_address_info(hostname, port)
+        sock = _try_create_socket_from_address_info(socket_info[0], use_insecure, certs)
+        return sock
+
+    def _get_socket_address_info(self, hostname, port):
+        info = socket.getaddrinfo(hostname, port, 0, self.socktype)
+        if not info:
+            raise OSError("getaddrinfo() returns an empty list")
+        return info
+
+    def emit(self, record):
+        try:
+            self._send_record(record)
+        except Exception:
+            self.handleError(record)
+
+    def _send_record(self, record):
+        formatted_record = self.format(record)
+        msg = formatted_record + "\n"
+        msg = msg.encode("utf-8")
+        if self.socktype == socket.SOCK_DGRAM:
+            self.socket.sendto(msg, self.address)
+        else:
+            self.socket.sendall(msg)
+
+    def close(self):
+        self.socket.close()
+        logging.Handler.close(self)
+
+
+def _try_create_socket_from_address_info(info, use_insecure, certs):
+    address_family, sock_type, proto, _, sa = info
+    sock = None
+    try:
+        sock = _create_socket_from_uncoupled_address_info(
+            address_family, sock_type, proto, use_insecure, certs, sa
+        )
+    except Exception as exc:
+        if sock is not None:
+            sock.close()
+        raise exc
+
+    return sock
+
+
+def _create_socket_from_uncoupled_address_info(
+    address_family, sock_type, proto, use_insecure, certs, sa
+):
+    sock = socket.socket(address_family, sock_type, proto)
+    if not use_insecure:
+        sock = _wrap_socket_for_ssl(sock, certs)
+    if sock_type == socket.SOCK_STREAM:
+        sock = _connect_socket(sock, sa)
+    return sock
+
+
+def _wrap_socket_for_ssl(sock, certs):
+    certs = certs or None
+    cert_reqs = ssl.CERT_REQUIRED if certs else ssl.CERT_NONE
```

Comment:
This is not what I would expect the default behavior to be. If I don't provide a cert file, it doesn't do any validation? If an end-user's server has a valid CA-signed cert, then they would expect to not need to provide a cert file, and would expect it to validate automatically.

I think this should be: If I try to connect to a server with an invalid/self-signed cert without using `--cert` option, it should always warn and fail. I can use `--cert` to provide that cert if I know it's valid but just not CA-signed. And we should maybe have a way to explicitly skip validation (maybe `--certs ignore`?), for those cases where it's a self-signed cert on the server, but they don't have access to that cert file. 

We can do this by something like:
```
context = ssl.create_default_context(cafile=certs, server_hostname=host)
sock = context.wrap_socket(sock)
```

If `cafile` is None, it defaults to using OS cert chains for validation. 

https://docs.python.org/3/library/ssl.html#socket-creation

Reply:
Ah, that makes sense. I did not consider automatic cert validation.

Is an inaccessible self-signed cert the only case where skipping validation would make sense? Seems like the user should just use TCP alone then.

Reply:
I want to make sure of this behavior, because the ssl.wrap_context method defaults to CERT_NONE if you don't pass any certs

Reply:
Have you confirmed that automatica cert validation is not currently happening?

Reply:
That's probably the main case, but there could be others. Like if the server had a valid CA-signed cert, but dns lookup wasn't working on the client machine so they had to use just the host IP.. 

But plain TCP wouldn't work in the cases where the server is only listening for TLS connections, as the connection would get rejected server-side.

Reply:
Ah! Looks like you are right. This top level method is actually deprecated for reasons similar to this.

Reply:
Is the phrase `CERT_NONE` common enough that we could have `--certs NONE` or `--certs none` work for ignoring them?

Reply:
CURL uses `--insecure` in this case, so we could have an extra flag that only applies to `TLS`, but I can see that getting confusing

Reply:
I'd rather have the meaning explicit in what the user has to type, so `--certs none` doesn't quite do it. An additional `--insecure` option is fine IMO.

Reply:
`--insecure` only applies to certain types of request in curl too 😛 . I agree with @timabrmsn that being insecure is something we would want to make sure someone doesn't opt into unknowingly, we should be as clear as possible here.

Reply:
@timabrmsn good find in considering certs signed by a valid CA. That wasn't something that I hadn't considered since this functionality is typically used to point at an on-prem syslog server, but it's certainly a valid case.

Reply:
The reason I thought it would confusing was because of the UDP/TCP - those could also be considered insecure.

Perhaps `--ignore-cert-validation` is best. Explicit.

Reply:
Similarly to before, I can't ever successfully use certs, maybe because they are self-signed?
Other than that, I think the behavior of using the `--ignore-cert-validation` works now the same as it did before when excluding the `--certs` flag.  Perhaps there are some implementation improvements yet though

Reply:
I got it to work by generating a key + cert with localhost and 127.0.0.1 as subject alternate name (SAN) params (https://medium.com/@antelle/how-to-generate-a-self-signed-ssl-certificate-for-an-ip-address-f0dd8dddf754)

Create a config file:
```
[req]
default_bits  = 2048
distinguished_name = req_distinguished_name
req_extensions = req_ext
x509_extensions = v3_req
prompt = no
[req_distinguished_name]
countryName = XX
stateOrProvinceName = N/A
localityName = N/A
organizationName = Self-signed certificate
commonName = 120.0.0.1: Self-signed certificate
[req_ext]
subjectAltName = @alt_names
[v3_req]
subjectAltName = @alt_names
[alt_names]
IP.1 = 127.0.0.1
DNS.1 = localhost
```

Generate the cert+key from the config: 

```
openssl req -x509 -nodes -days 730 -newkey rsa:2048 -keyout key.pem -out cert.pem -config san.cnf
```

Now tell ncat to use that cert when listening with ssl: 

```
ncat -kl -p 9999 --ssl --ssl-cert cert.pem --ssl-key key.pem
```

Running the CLI without --certs gives us the self-signed error:

```
code42 audit-logs send-to localhost:9999 -b 1d -p TLS-TCP 
Error: Unable to connect to localhost:9999. Failed with error: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: self signed certificate (_ssl.c:1122).
```

Providing the same cert ncat is using works: 

```
code42 audit-logs send-to localhost:9999 -b 1d -p TLS-TCP --certs cert.pem
```

Reply:
heck yeah, thanks for these instructions